brainiak/matnormal/covs.py

```diff
@@ -511,66 +533,79 @@ def __init__(self, sizes, Sigmas=None, mask=None):
         self.nfactors = len(sizes)
         self.size = np.prod(np.array(sizes), dtype=np.int32)
 
+        npar = [(size * (size + 1)) // 2 for size in self.sizes]
         if Sigmas is None:
-            self.L_full = [
+            self.Lflat = [
                 tf.Variable(
-                    tf.random.normal([sizes[i], sizes[i]], dtype=tf.float64),
-                    name="L" + str(i) + "_full",
+                    tf.random.normal([npar[i]], dtype=tf.float64),
+                    name="L" + str(i) + "_flat",
                 )
                 for i in range(self.nfactors)
             ]
         else:
-            self.L_full = [
-                tf.Variable(np.linalg.cholesky(
-                    Sigmas[i]), name="L" + str(i) + "_full")
+            self.Lflat = [
+                tf.Variable(
+                    flatten_cholesky_unique(np.linalg.cholesky(Sigmas[i])),
+                    name="L" + str(i) + "_flat",
+                )
                 for i in range(self.nfactors)
             ]
         self.mask = mask
 
-        # make a list of choleskys
-        L_indeterminate = [tf.linalg.band_part(
-            mat, -1, 0) for mat in self.L_full]
-        self.L = [
-            tf.linalg.set_diag(mat, tf.exp(tf.linalg.diag_part(mat)))
-            for mat in L_indeterminate
-        ]
+    @property
+    def L(self):
+        return [unflatten_cholesky_unique(mat) for mat in self.Lflat]
 
     def get_optimize_vars(self):
         """ Returns a list of tf variables that need to get optimized
             to fit this covariance
         """
-        return self.L_full
+        return self.Lflat
 
     @property
     def logdet(self):
         """ log|Sigma| using the diagonals of the cholesky factors.
         """
         if self.mask is None:
-            n_list = tf.stack([tf.cast(tf.shape(input=mat)[0], dtype=tf.float64)
-                               for mat in self.L])
+            n_list = tf.stack(
+                [tf.cast(tf.shape(input=mat)[0], dtype=tf.float64)
+                 for mat in self.L]
+            )
             n_prod = tf.reduce_prod(input_tensor=n_list)
             logdet = tf.stack(
-                [tf.reduce_sum(input_tensor=tf.math.log(tf.linalg.tensor_diag_part(mat)))
-                 for mat in self.L]
+                [
+                    tf.reduce_sum(
+                        input_tensor=tf.math.log(
+                            tf.linalg.tensor_diag_part(mat))
```

Comment:
same concern, if the one above is valid.

Reply:
Also defined via unique cholesky parameterization, should be valid.